### ISSUE TITLE
Add WebAssembly System Interface (WASI) ABI support

### ISSUE THREAD

**Issue by Uzume** [WAVM/WAVM#155]
It would be nice it WAVM had [WASI ](https://github.com/WebAssembly/WASI) ABI support as several other standalone WebAssembly VMs already do.

**Comment by AndrewScheidecker** [comment]
WAVM does have rudimentary WASI support, though only for command-line arguments, clocks, and stdin/stdout/stderr I/O. Use the `wavm-run-wasi` command to run a WASI program.

**Comment by Uzume** [comment]
Yes, I see some work was done in [/Lib/WASI/WASI.cpp](https://github.com/WAVM/WAVM/blob/master/Lib/WASI/WASI.cpp)  and the entire API has been stubbed out but many are still unimplemented and I imagine few have had any serious testing.

**Comment by AndrewScheidecker** [comment]
You're right, there's a lot of work left to do! :)

**Comment by duanbing** [comment]
Hi, Andrew:
    Maybe you can list a todo of supporting WASI, so we also can do some
contribution for it.

Andrew Scheidecker <notifications@github.com> 于2019年5月16日周四 上午5:13写道：

**Comment by AndrewScheidecker** [comment]
WAVM currently supports most of the WASI API. These API entry points are not yet implemented:
* `poll_oneoff`
* `proc_raise` (likely to be removed from WASI)
* `sock_recv`
* `sock_send`
* `sock_shutdown`
* `fd_allocate`
* `fd_readdir`
* `path_link`
* `path_readlink`
* `path_rename`
* `path_symlink`

**Comment by laothuong** [comment]
> VMs

**Comment by vshymanskyy** [comment]
`path_rename` prevents us from running `clang.wasm`:
https://github.com/wasm3/wasm3-self-compiling/blob/7a251d9483cdd4f559e7cf4af36a6468af100ea4/Makefile#L28-L32

Also, when running `wasm-ld.wasm`, I get "lld is a generic driver".
Probably, there is some issue in the `args_get` function.

**Comment by AndrewScheidecker** [comment]
@vshymanskyy I implemented `path_rename`, but the clang.wasm from that repo is still failing on WAVM for some reason I haven't figured out yet. Thanks for the test case.

**Comment by vshymanskyy** [comment]
@AndrewScheidecker I think the reason for this failure is that `wasi_unstable` differs from `wasi_snapshot_preview1` a bit, primarily in structure layout.
Properly implementing `fd_filestat_get`, `path_filestat_get`, `fd_seek` should be enough for `clang.wasm`.
